experimental/sql/src/projection_optimizations.py

```diff
@@ -0,0 +1,191 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import (RewritePattern, GreedyRewritePatternApplier,
+                                   PatternRewriteWalker, PatternRewriter,
+                                   op_type_rewrite_pattern)
+
+import dialects.rel_alg as RelAlg
+
+# This file contains several rewrites on the rel_alg level, all of which concern
+# projections. Currently, there are three rewriters. The first one introduces a
+# projection before any operator, the second one fuses subsequent projections,
+# and the third one removes projetions that are just identity mappings.
+
+
+@dataclass
+class ProjectionOptimizer(RewritePattern):
+
+  def flatten(self, l: list[list[str]]) -> list[str]:
+    return [elem for sublist in l for elem in sublist]
+
+  def find_cols_in_expr(self, op: RelAlg.Expression) -> list[str]:
+    """
+    Returns the list of cols that are used in this expression or its children.
+    """
+    if isinstance(op, RelAlg.Column):
+      return [op.col_name.data]
+    else:
+      return self.flatten([self.find_cols_in_expr(r.op) for r in op.regions])
+
+  def find_cols_upstream(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its parents. If
+    the list contains None, a ModuleOp (so the output) was hit, showing that all
+    columns need to be kept.
+    """
+    if isinstance(op, ModuleOp):
+      return [None]
+    if isinstance(op, RelAlg.Select):
```

Comment:
It may not be possible here but I'd like to be sure: can we remove the type-switch here and use a more generic traversal?

Reply:
I think it might be possible to write this with less cases, but I feel that would make a lot of assumptions on the general structure of rel_alg, so I would prefer to not do that.

Notice that the recursive call is only called in some cases, sometimes we need a call to `find_cols_in_expr`, some times we don't...

Reply:
I am not so much worried about the recursion than I am about splitting the knowledge about each individual op across many different places in the code base. With such a design, you need to touch many many places every time you add a new op and you can easily forgot some. (On the other hand, maybe this won't happen too often for relational algebra ;-) And of course, it is only possible if you can guarantee correctness based on common properties/assumptions...)